src/storage/src/table/state_table.rs

```diff
@@ -102,28 +111,137 @@ impl<S: StateStore> StateTable<S> {
+        let state_table_iter = Self {
+            mem_table_iter,
+            cell_based_streaming_iter,
+            cell_based_item,
+            pk_serializer,
+        };
+
+        Ok(state_table_iter)
     }
 
-    async fn next(&mut self) -> StorageResult<Option<Row>> {
-        todo!()
+    pub async fn next(&mut self) -> StorageResult<Option<Row>> {
+        let next_flag;
+        let mut res = None;
+        let cell_based_item = self.cell_based_item.take();
+        match (cell_based_item, self.mem_table_iter.peek()) {
+            (None, None) => {
+                next_flag = NextOutcome::Neither;
+                res = None;
+            }
+            (Some((_, row)), None) => {
+                res = Some(row);
+                next_flag = NextOutcome::Storage;
+            }
+            (None, Some((_, row_op))) => {
+                next_flag = NextOutcome::MemTable;
+                match row_op {
+                    RowOp::Insert(row) => res = Some(row.clone()),
+                    RowOp::Delete(_) => {}
+                    RowOp::Update((_, new_row)) => res = Some(new_row.clone()),
+                }
+            }
+            (Some((cell_based_pk, cell_based_row)), Some((mem_table_pk, mem_table_row_op))) => {
+                let mem_table_pk_bytes =
+                    serialize_pk(mem_table_pk, &self.pk_serializer).map_err(err)?;
+                match cell_based_pk.cmp(&mem_table_pk_bytes) {
+                    Ordering::Less => {
+                        // cell_based_table_item will be return
+                        res = Some(cell_based_row);
+                        next_flag = NextOutcome::Storage;
+                    }
+                    Ordering::Equal => {
+                        // mem_table_item will be return, while both cell_based_streaming_iter and
+                        // mem_table_iter need to execute next() once.
+
+                        next_flag = NextOutcome::Both;
+                        match mem_table_row_op {
+                            RowOp::Insert(row) => {
+                                res = Some(row.clone());
+                            }
+                            RowOp::Delete(_) => {}
+                            RowOp::Update((old_row, new_row)) => {
+                                debug_assert!(old_row == &cell_based_row);
+                                res = Some(new_row.clone());
+                            }
+                        }
+                    }
+                    Ordering::Greater => {
+                        // mem_table_item will be return
+                        next_flag = NextOutcome::MemTable;
+
+                        match mem_table_row_op {
+                            RowOp::Insert(row) => res = Some(row.clone()),
+                            RowOp::Delete(_) => {}
+                            RowOp::Update((_, new_row)) => {
+                                res = Some(new_row.clone());
```

Comment:
I suddenly realize that this branch will never be hit. If we have update, there must be a record in shared storage. We may simply panic here.

Reply:
So it is.